### PR TITLE
Retry on response format failure

### LINKED ISSUE #86: Retry when structured output fails
Thing things we need to check and then retry if they fail
(1) the response is valid JSON (try-catch current in `parse_response_message`) 
(2) the response JSON matches the Pydantic model schema (base_request_processor in `create_dataset_files`) 

Right now these checks happen after the LLM responds (successfully sends back a string) so we don't have it within the retry logic. For batch, we would need to accumulate all the failed responses and send a new batch. 
For Online, we can add these checks before writing to `responses.jsonl` file. However the logic might be weird since everything about this parsing was abstracted away from the requests --> responses logic in online. Might need to rethink the pathways a little bit. 

Lessons learned from #85

### PR DESCRIPTION
**Pull request by RyanMarten**
Fixes #86

### CODE REVIEW

**Comment by RyanMarten** [comment]
As discussed in #86, using `strict: true` could help. This argument make sense to expose through the context manager that Mahesh was working on. 
It only an available argument for some models (subset of openai models). Should we verify ourselves with nice errors about supported param? Litellm does that. Or do we have a more sophisticated retry logic (based off tenacity - as Mahesh suggested and noted in #261 ) that doesn't retry when the response code says the request shape is not value.

**Comment by RyanMarten** [comment]
Looking for examples for a minimal reproduction. @neginraoof can maybe provide. 

In https://rwilinski.ai/posts/benchmarking-llms-for-structured-json-generation/ the gpt-4o-mini gets 100% on all the attempts.

**Comment by RyanMarten** [comment]
Had claude help come up with an example


Which the model struggles at creating a valid response to

```
from bespokelabs.curator import LLM
from datasets import Dataset
from pydantic import BaseModel, Field
from typing import List, Dict, Optional
from enum import Enum
from datetime import datetime

# Complex dataset with tricky prompts
dataset = Dataset.from_dict({
    "prompt": [
        """Analyze this fictional company's performance and provide a detailed breakdown:
        TechCorp Industries faced challenges in Q3 2023. Their cloud division struggled 
        while IoT solutions showed promise. Customer satisfaction varied by region."""
    ]
})

class SentimentLevel(str, Enum):
    VERY_NEGATIVE = "very_negative"
    NEGATIVE = "negative"
    NEUTRAL = "neutral"
    POSITIVE = "positive"
    VERY_POSITIVE = "very_positive"

class MetricScore(BaseModel):
    value: float
    confidence_interval: List[float] = Field(min_items=2, max_items=2)
    reliability_score: float = Field(ge=0, le=1)
    last_updated: datetime

class DepartmentAnalysis(BaseModel):
    department_name: str
    revenue_change: float
    key_metrics: Dict[str, MetricScore]
    risk_factors: List[Dict[str, float]]
    sentiment: SentimentLevel
    regional_breakdown: Dict[str, Dict[str, float]]
    future_projections: List[Dict[str, Optional[float]]]

class CompanyAnalysis(BaseModel):
    company_overview: Dict[str, str]
    financial_health: float = Field(ge=-1, le=1)
    departments: List[DepartmentAnalysis]
    market_position: Dict[str, List[float]]
    competitor_comparison: Dict[str, Dict[str, float]]
    risk_assessment: List[Dict[str, Optional[float]]]
    timestamp: datetime
    confidence_score: float = Field(ge=0, le=1)

prompter = LLM(
    prompt_func=lambda row: row["prompt"],
    model_name="gpt-4o-mini",
    response_format=CompanyAnalysis,
)

dataset = prompter(dataset)
print(dataset.to_pandas())
```

File: src/bespokelabs/curator/request_processor/base_request_processor.py [hunk 1]
```diff
@@ -262,6 +262,58 @@ def attempt_loading_cached_dataset(
                     "Deleted file and attempting to regenerate dataset from cached LLM responses."
                 )
 
+    def convert_response_to_response_format(
+        self, response_message: str | dict, response_format: Optional[BaseModel]
+    ) -> Optional[dict | str]:
+        """
+        Converts a response message to a specified Pydantic model format.
```
Comment (madiator) [inline src/bespokelabs/curator/request_processor/base_request_processor.py:269]: A bit lost on the motivation here.

Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_request_processor.py:269]: this function we use anyways, I just moved it out so we don't duplicate code

We need to convert to the correct response format from the string that model returns

Comment (madiator) [reply src/bespokelabs/curator/request_processor/base_request_processor.py:269]: Ah sorry.
Also, does it make sense to put this in prompt_formatter?
So `self.prompt_formatter.convert_response_to_response_format(response_message)`.

**Review by madiator — APPROVED** [review]

**Pull request merged**